src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/SnapshotTrackerTransformer.kt

```diff
@@ -0,0 +1,235 @@
+    private val isInstanceOf: (actualType: String, expectedSuperType: String) -> Boolean
+) : ManagedStrategyWithAnalyzerClassVisitor(fileName, className, methodName, adapter) {
+
+    /**
+     * Searches for invocations of constructors `className::<init>(args)` and inserts bytecode
+     * that will extract all objects from `args` which types are subtypes of the class `className`.
+     * Snapshot tracker then tracks the extracted objects energetically.
+     *
+     * This is a hack solution to the problem of impossibility of identifying whether
+     * some object on stack in the constructor is a `this` reference or not.
+     * So for instructions `GETFIELD`/`PUTFIELD` in constructors handler of `visitFieldInsn(...)` method below
+     * we ignore the accesses to fields which owners are the same type as the constructor owner.
+     * The optimization is done to avoid 'leaking this' problem during bytecode verification.
+     * But sometimes `GETFIELD`/`PUTFIELD` are invoked on fields with owner objects are the same type as a constructor owner,
+     * but which are not `this`. This can happen when:
+     * - a non-static object is accessed which was passed to constructor arguments
+     * - a non-static object is accessed which was created locally in constructor
+     *
+     * We only case about the 1nd case, because for the 3nd case when dealing with
+     * a locally constructed object it can't be reached from static memory, so no need to track its fields.
+     * The 2nd case where such object is passed via constructor argument is handled here.
+     */
+    override fun visitMethodInsn(opcode: Int, owner: String, name: String, desc: String, itf: Boolean) = adapter.run {
+        if (name == "<init>") {
+            val matchedArguments = getArgumentTypes(desc).toList()
+                .mapIndexed { index, type ->
+                    if (
+                        !isArray(type) &&
+                        !isPrimitive(type) &&
+                        isInstanceOf(type.className.replace('.', '/'), owner)
+                    ) index
+                    else null
+                }
+                .filterNotNull()
+                .toIntArray()
+
+            if (matchedArguments.isEmpty()) {
+                visitMethodInsn(opcode, owner, name, desc, itf)
+                return
+            }
+
+            invokeIfInTestingCode(
+                original = { visitMethodInsn(opcode, owner, name, desc, itf) },
+                code = {
+                    // STACK: args
+                    val arguments = storeArguments(desc)
+                    val matchedLocals = arguments.filterIndexed { index, _ -> matchedArguments.contains(index) }
+                    // STACK: <empty>
+                    push(matchedLocals.size)
+                    // STACK: length
+                    visitTypeInsn(ANEWARRAY, "java/lang/Object")
+                    // STACK: array
+                    matchedLocals.forEachIndexed { index, local ->
```

Comment:
There is a similar function for this purpose in `TransformationUtils.kt` --- `storeLocals`, please use it instead.

Reply:
`pushArray(matchedLocals)` you probably meant
done